File: app/src/main/java/com/elli0tt/rpg_life/presentation/quests/QuestsFragment.java
package com.elli0tt.rpg_life.presentation.quests;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.view.ActionMode;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;
import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.elli0tt.rpg_life.R;
import com.google.android.material.floatingactionbutton.FloatingActionButton;


public class QuestsFragment extends Fragment {

    private QuestsViewModel viewModel;

    private QuestsAdapter questsAdapter = new QuestsAdapter();
    private NavController navController;

    private FloatingActionButton fab;
    private RecyclerView recyclerView;

    private ActionMode actionMode;

    private boolean isToShowDevelopersOptions;

    private MenuItem showCompletedMenuItem;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_quests, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        navController = NavHostFragment.findNavController(this);

        fab = view.findViewById(R.id.quests_fab);
        recyclerView = view.findViewById(R.id.quests_recycler_view);

        subscribeToViewModel();
        setHasOptionsMenu(true);
        setupQuestsRecyclerView();

        fab.setOnClickListener(v -> navigateToAddQuestScreen());
    }

    @Override
    public void onResume() {
        super.onResume();
        questsAdapter.notifyDataSetChanged();
        SharedPreferences sharedPreferences =
                PreferenceManager.getDefaultSharedPreferences(getContext());
        isToShowDevelopersOptions = sharedPreferences.getBoolean("developers_options", false);
        if (getActivity() != null) {
            getActivity().invalidateOptionsMenu();
        }
    }

    private QuestsAdapter.OnItemClickListener onItemClickListener = position -> {
        if (viewModel.getQuests().getValue() != null) {
            navigateToEditQuestScreen(viewModel.getQuests().getValue().get(position).getId());
        }
    };

    private void setupQuestsRecyclerView() {
        recyclerView.setAdapter(questsAdapter);
        questsAdapter.setOnItemClickListener(onItemClickListener);
        questsAdapter.setOnItemLongClickListener(position -> {
            AppCompatActivity activity = (AppCompatActivity) getActivity();
            if (activity != null) {
                actionMode = activity.startSupportActionMode(
                        new ActionModeController(viewModel, questsAdapter));
            }
            viewModel.startSelection();
            questsAdapter.startSelection(position);
        });
        questsAdapter.setOnSelectionFinishedListener(() -> {
            if (actionMode != null) {
                actionMode.finish();
                actionMode = null;
            }
        });
        questsAdapter.setOnIsCompleteCheckBoxClickListener((isCompleted, position) ->
                viewModel.completeQuest(position, isCompleted));
        questsAdapter.setOnIsImportantCheckBoxClickListener((isImportant, position) ->
                viewModel.setQuestImportant(position, isImportant));

        recyclerView.setLayoutManager(new LinearLayoutManager(getContext(), RecyclerView.VERTICAL
                , false));
//        recyclerView.addItemDecoration(new DividerItemDecoration(
//                Objects.requireNonNull(getContext()), RecyclerView.VERTICAL));
        recyclerView.addItemDecoration(new DividerItemDecoration(getContext(),
                DividerItemDecoration.VERTICAL));

        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                if (dy < 0 && !fab.isShown() && !viewModel.isSelectionStarted().getValue()) {
                    fab.show();
                } else if (dy > 0 && fab.isShown()) {
                    fab.hide();
                }
            }
        });
    }

    private void subscribeToViewModel() {
        viewModel = ViewModelProviders.of(this).get(QuestsViewModel.class);
        viewModel.getQuests().observe(getViewLifecycleOwner(), questList -> {
            questsAdapter.submitList(questList);
        });
        viewModel.isSelectionStarted().observe(getViewLifecycleOwner(), new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean isSelectionStarted) {
                if (isSelectionStarted) {
                    fab.hide();
                } else {
                    fab.show();
                }
            }
        });
        viewModel.getShowCompletedTextResId().observe(getViewLifecycleOwner(),
                new Observer<Integer>() {
                    @Override
                    public void onChanged(Integer textResId) {
                        if (showCompletedMenuItem != null) {
                            showCompletedMenuItem.setTitle(textResId);
                        }
                    }
                });
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.quests_toolbar_menu, menu);

        MenuItem populateWithSamplesItem =
                menu.findItem(R.id.quests_toolbar_menu_populate_with_samples);
        MenuItem deleteAllItem = menu.findItem(R.id.quests_toolbar_menu_delete_all);
        showCompletedMenuItem = menu.findItem(R.id.quests_toolbar_menu_show_completed);

        populateWithSamplesItem.setVisible(isToShowDevelopersOptions);
        deleteAllItem.setVisible(isToShowDevelopersOptions);
        showCompletedMenuItem.setTitle(viewModel.getShowCompletedTextResId().getValue());
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.quests_toolbar_menu_delete_all:
                viewModel.deleteAll();
                fab.show();
                break;
            case R.id.quests_toolbar_menu_populate_with_samples:
                viewModel.populateWithSamples();
                break;
            case R.id.quests_toolbar_menu_sort_by_name:
                viewModel.setSorting(QuestsSortingState.NAME);
                break;
            case R.id.quests_toolbar_menu_sort_by_date_due:
                viewModel.setSorting(QuestsSortingState.DATE_DUE);
                break;
            case R.id.quests_toolbar_menu_sort_by_date_added:
                viewModel.setSorting(QuestsSortingState.DATE_ADDED);
                break;
            case R.id.quests_toolbar_menu_sort_by_difficulty:
                viewModel.setSorting(QuestsSortingState.DIFFICULTY);
                break;
            case R.id.quests_toolbar_menu_filtering_all:
                viewModel.setFiltering(QuestsFilterState.ALL);
                break;
            case R.id.quests_toolbar_menu_filtering_important:
                viewModel.setFiltering(QuestsFilterState.IMPORTANT);
                break;
            case R.id.quests_toolbar_menu_filter_today:
                viewModel.setFiltering(QuestsFilterState.TODAY);
                break;
            case R.id.quests_toolbar_menu_filter_tomorrow:
                viewModel.setFiltering(QuestsFilterState.TOMORROW);
                break;
            case R.id.quests_toolbar_menu_show_completed:
                viewModel.changeShowCompleted();
        }
        return true;
    }

    private void navigateToAddQuestScreen() {
        QuestsFragmentDirections.ActionQuestsScreenToAddQuestScreen action =
                QuestsFragmentDirections.actionQuestsScreenToAddQuestScreen();
        navController.navigate(action);
    }

    private void navigateToEditQuestScreen(int questId) {
        QuestsFragmentDirections.ActionQuestsScreenToEditQuestScreen action =
                QuestsFragmentDirections.actionQuestsScreenToEditQuestScreen();
        action.setQuestId(questId);
        navController.navigate(action);
    }
}

File: app/src/main/java/com/elli0tt/rpg_life/presentation/quests/QuestsFragment.kt
package com.elli0tt.rpg_life.presentation.quests

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.os.Bundle
import android.util.Log
import android.view.*
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.ActionMode
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.NavController
import androidx.navigation.fragment.NavHostFragment
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.work.WorkInfo
import com.elli0tt.rpg_life.R
import com.elli0tt.rpg_life.domain.model.Quest
import com.google.android.material.floatingactionbutton.FloatingActionButton

class QuestsFragment : Fragment() {
    private lateinit var viewModel: QuestsViewModel
    private val questsAdapter = QuestsAdapter()

    private lateinit var navController: NavController
    private lateinit var addQuestFab: FloatingActionButton
    private lateinit var addChallengeFab: FloatingActionButton
    private lateinit var mainFab: FloatingActionButton
    private lateinit var recyclerView: RecyclerView
    private lateinit var fabMenuBackgroundView: View
    private lateinit var addChallengeCardView: View
    private lateinit var addQuestCardView: View

    private var actionMode: ActionMode? = null
    private var isToShowDevelopersOptions = false
    private var showCompletedMenuItem: MenuItem? = null

    private var isFabMenuOpened = false

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_quests, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        navController = NavHostFragment.findNavController(this)

        addQuestFab = view.findViewById(R.id.add_quest_fab)
        addChallengeFab = view.findViewById(R.id.add_challenge_fab)
        mainFab = view.findViewById(R.id.main_fab)
        recyclerView = view.findViewById(R.id.quests_recycler_view)
        fabMenuBackgroundView = view.findViewById(R.id.fab_menu_background_view)
        addChallengeCardView = view.findViewById(R.id.add_challenge_card_view)
        addQuestCardView = view.findViewById(R.id.add_quest_card_view)

        subscribeToViewModel()
        setHasOptionsMenu(true)
        setupQuestsRecyclerView()

        addQuestFab.setOnClickListener {
            hideFabMenu()

            viewModel.insertEmptyQuest()
        }
        addChallengeFab.setOnClickListener {
            hideFabMenu()
            navigateToAddChallengeScreen()
        }
        mainFab.setOnClickListener {
            if (!isFabMenuOpened) {
                showFabMenu()
            } else {
                hideFabMenu()
            }
        }
        fabMenuBackgroundView.setOnClickListener {
            hideFabMenu()
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        (activity as AppCompatActivity).supportActionBar?.setDisplayHomeAsUpEnabled(false)
    }

    override fun onResume() {
        super.onResume()
        questsAdapter.notifyDataSetChanged()
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
        isToShowDevelopersOptions = sharedPreferences.getBoolean("developers_options", false)
        if (activity != null) {
            requireActivity().invalidateOptionsMenu()
        }
    }

    private val onItemClickListener = QuestsAdapter.OnItemClickListener { position: Int ->
        val quests = viewModel.quests.value
        if (quests != null) {
            if (quests[position].isChallenge) {
                navigateToEditChallengeScreen(quests[position].id)
            } else {
                navigateToEditQuestScreen(viewModel.quests.value!![position].id)
            }
        }
    }

    private fun setupQuestsRecyclerView() {
        recyclerView.adapter = questsAdapter
        questsAdapter.setOnItemClickListener(onItemClickListener)
        questsAdapter.setOnItemLongClickListener { position: Int ->
            val activity: AppCompatActivity? = activity as AppCompatActivity?
            if (activity != null) {
                actionMode = activity.startSupportActionMode(
                        ActionModeController(viewModel, questsAdapter))
            }
            viewModel.startSelection()
            questsAdapter.startSelection(position)
        }
        questsAdapter.setOnSelectionFinishedListener {
            if (actionMode != null) {
                actionMode!!.finish()
                actionMode = null
            }
        }

        questsAdapter.setOnIsCompleteCheckBoxClickListener { isCompleted: Boolean, position: Int ->
            viewModel.completeQuest(position, isCompleted)
        }
        questsAdapter.setOnIsImportantCheckBoxClickListener { isImportant: Boolean, position: Int ->
            viewModel.setQuestImportant(position, isImportant)
        }
        recyclerView.layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                if ((dy < 0) && !mainFab.isShown && !viewModel.isSelectionStarted.value!!) {
                    mainFab.show()
                } else if (dy > 0 && addQuestFab.isShown) {
                    mainFab.hide()
                }
            }
        })
    }

    private fun subscribeToViewModel() {
        viewModel = ViewModelProvider(this).get(QuestsViewModel::class.java)
        viewModel.quests.observe(viewLifecycleOwner, Observer { questList: List<Quest?>? -> questsAdapter.submitList(questList) })
        viewModel.isSelectionStarted.observe(viewLifecycleOwner, Observer { isSelectionStarted ->
            if (isSelectionStarted) {
                mainFab.hide()
            } else {
                mainFab.show()
            }
        })
        viewModel.showCompletedTextResId.observe(viewLifecycleOwner, Observer { textResId: Int? ->
            if (showCompletedMenuItem != null) {
                showCompletedMenuItem!!.setTitle((textResId)!!)
            }
        })
        viewModel.workInfo.observe(viewLifecycleOwner, workInfoObserver)
    }

    private val workInfoObserver = object : Observer<WorkInfo>{
        override fun onChanged(workInfo: WorkInfo?) {
            if (workInfo != null && workInfo.state.isFinished){
                navigateToEditQuestScreen(workInfo.outputData.getInt(Constants.KEY_QUEST_ID, 0))
                viewModel.updateInsertWorkRequest()
            }
        }
    }

    private fun showFabMenu() {
        isFabMenuOpened = true
        fabMenuBackgroundView.visibility = View.VISIBLE
        addQuestFab.show()
        addChallengeFab.show()

        mainFab.animate().rotation(135f)
        fabMenuBackgroundView.animate().alpha(1f)
        addChallengeFab.animate()
                .translationY(-requireContext().resources.getDimension(R.dimen.add_challenge_fab_translationY))
                .rotation(0f)
        addQuestFab.animate()
                .translationY(-requireContext().resources.getDimension(R.dimen.add_quest_fab_translationY))
                .rotation(0f)
                .setListener(object : AnimatorListenerAdapter() {
                    override fun onAnimationEnd(animation: Animator?) {
                        super.onAnimationEnd(animation)
                        addQuestCardView.visibility = View.VISIBLE
                        addChallengeCardView.visibility = View.VISIBLE
                    }
                })
    }

    private fun hideFabMenu() {
        isFabMenuOpened = false
        mainFab.animate().rotation(0f)
        fabMenuBackgroundView.animate().alpha(0f)
        addChallengeFab.animate()
                .translationY(0f)
                .rotation(90f)
        addQuestFab.animate()
                .translationY(0f)
                .rotation(90f)
                .setListener(object : AnimatorListenerAdapter() {
                    override fun onAnimationEnd(animation: Animator?) {
                        super.onAnimationEnd(animation)
                        addQuestCardView.visibility = View.GONE
                        addChallengeCardView.visibility = View.GONE
                        addChallengeFab.hide()
                        addQuestFab.hide()
                        fabMenuBackgroundView.visibility = View.GONE
                    }
                })
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.quests_toolbar_menu, menu)
        val populateWithSamplesItem = menu.findItem(R.id.populate_with_samples)
        val deleteAllItem = menu.findItem(R.id.quests_toolbar_menu_delete_all)
        showCompletedMenuItem = menu.findItem(R.id.show_completed)
        populateWithSamplesItem.isVisible = isToShowDevelopersOptions
        deleteAllItem.isVisible = isToShowDevelopersOptions
        showCompletedMenuItem?.setTitle((viewModel.showCompletedTextResId.value)!!)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.quests_toolbar_menu_delete_all -> {
                viewModel.deleteAll()
                addQuestFab.show()
            }
            R.id.populate_with_samples -> viewModel.populateWithSamples()
            R.id.show_completed -> viewModel.changeShowCompleted()

            R.id.sort_by_name -> viewModel.setSorting(QuestsSortingState.NAME)
            R.id.sort_by_date_due -> viewModel.setSorting(QuestsSortingState.DATE_DUE)
            R.id.sort_by_date_added -> viewModel.setSorting(QuestsSortingState.DATE_ADDED)
            R.id.sort_by_difficulty -> viewModel.setSorting(QuestsSortingState.DIFFICULTY)

            R.id.filtering_all -> viewModel.setFiltering(QuestsFilterState.ALL)
            R.id.filtering_important -> viewModel.setFiltering(QuestsFilterState.IMPORTANT)
            R.id.filter_today -> viewModel.setFiltering(QuestsFilterState.TODAY)
            R.id.filter_tomorrow -> viewModel.setFiltering(QuestsFilterState.TOMORROW)
        }
        return true
    }

    private fun navigateToAddQuestScreen() {
        val action = QuestsFragmentDirections.actionQuestsScreenToAddQuestScreen()
        navController.navigate(action)
    }

    private fun navigateToEditQuestScreen(questId: Int) {
        val action = QuestsFragmentDirections.actionQuestsScreenToEditQuestScreen()
        action.questId = questId
        navController.navigate(action)
    }

    private fun navigateToAddChallengeScreen() {
        val action = QuestsFragmentDirections.actionQuestsScreenToAddEditChallengeScreen()
        navController.navigate(action)
    }

    private fun navigateToEditChallengeScreen(challengeId: Int) {
        val action = QuestsFragmentDirections.actionQuestsScreenToAddEditChallengeScreen()
        action.challengeId = challengeId
        navController.navigate(action)
    }
}